apps/server/src/modules/task/uc/task.uc.spec.ts

```diff
@@ -91,449 +70,401 @@ describe('TaskUC', () => {
-			user = setupUser(permissions);
-		});
-
-		it('should call auth getUserWithPermissions() with userId', async () => {
-			const mockRestore = findAllMock();
-
-			await service.findAllFinished(user.id);
+		describe('when user has read permission', () => {
+			describe('When task has user parent and is found successfully', () => {
+				const setup = () => {
+					const permissions = [Permission.TASK_DASHBOARD_VIEW_V3];
+					const user = setupUser(permissions);
+					const task = taskFactory.finished(user).build();
 
-			expect(authorizationService.getUserWithPermissions).toHaveBeenCalledWith(user.id);
+					authorizationService.getUserWithPermissions.mockResolvedValueOnce(user);
+					courseRepo.findAllByUserId.mockResolvedValueOnce([[], 0]);
+					lessonRepo.findAllByCourseIds.mockResolvedValueOnce([[], 0]);
+					lessonRepo.findAllByCourseIds.mockResolvedValueOnce([[], 0]);
+					authorizationService.hasPermission.mockReturnValueOnce(false);
+					taskRepo.findAllFinishedByParentIds.mockResolvedValueOnce([[task], 1]);
 
-			mockRestore();
-		});
-
-		it('should return task for a user', async () => {
-			const task = taskFactory.finished(user).build();
-			const mockRestore = findAllMock({ tasks: [task] });
-
-			const [, total] = await service.findAllFinished(user.id);
-
-			expect(total).toEqual(1);
-
-			mockRestore();
-		});
-
-		it('should call task repo findAllFinishedByParentIds', async () => {
-			const mockRestore = findAllMock({});
-			const spy = setTaskRepoMock.findAllFinishedByParentIds();
+					return { user, task };
+				};
 
-			await service.findAllFinished(user.id);
+				it('should call auth getUserWithPermissions() with userId', async () => {
+					const { user } = setup();
 
-			expect(spy).toHaveBeenCalled();
+					await service.findAllFinished(user.id);
 
-			mockRestore();
-		});
-
-		it('should call task repo findAllFinishedByParentIds for finished tasks', async () => {
-			const mockRestore = findAllMock({});
-			const spy = setTaskRepoMock.findAllFinishedByParentIds();
+					expect(authorizationService.getUserWithPermissions).toHaveBeenCalledWith(user.id);
+				});
 
-			await service.findAllFinished(user.id);
+				it('should return task for a user', async () => {
+					const { user } = setup();
 
-			const expectedParams = [
-				{
-					creatorId: user.id,
-					openCourseIds: [],
-					finishedCourseIds: [],
-					lessonIdsOfOpenCourses: [],
-					lessonIdsOfFinishedCourses: [],
-				},
-				{ pagination: undefined, order: { dueDate: SortOrder.desc } },
-			];
-			expect(spy).toHaveBeenCalledWith(...expectedParams);
+					const [, total] = await service.findAllFinished(user.id);
 
-			mockRestore();
-		});
-
-		it('should return a counted type', async () => {
-			const mockRestore = findAllMock({});
+					expect(total).toEqual(1);
+				});
 
-			const [data, count] = await service.findAllFinished(user.id);
+				it('should call task repo findAllFinishedByParentIds', async () => {
```

Comment:
I think it is sufficient to test the return value of the whole function and the repo calls should not be tested. Because on the one hand the return value would change if the calls were just missing or wrongly executed. On the other hand, quite hypothetically, if you would get the data somehow else, the test should not fail.

I'm thinking about this guideline here: "stick to blackbox testing: think about the unit from the outside, not its inner workings" from https://hpi-schul-cloud.github.io/schulcloud-server/additional-documentation/nestjs-application/testing.html

Reply:
Not in focus here. Moved to miro board as todo